### PR TITLE
Optimize submission of tools that output collections.

### PR DESCRIPTION
**Pull request by jmchilton**
Some output timings for a tool that creates a pair vs a tool that creates two outputs. If you squint it looks like this branch reduces the delta between the two tools. Optimizing the individual parts of the code worked really well (logs statements not shown) but the final commutative product is a bit harder to see. I expected a bigger difference given the reported problems - might be worth taking another pass in the context of a mapping operation over a tool that produces pairs, this pass focused on executions that produce HDCAs which are difference than producing collections.

```
DEV

galaxy.tools.execute DEBUG 2020-05-29 18:35:29,208 [p:45925,w:1,m:0] [uWSGIWorker1Core2] Executed 1 job(s) for tool collection_creates_pair request (748.437 ms)

galaxy.tools.execute DEBUG 2020-05-29 18:36:00,763 [p:45925,w:1,m:0] [uWSGIWorker1Core0] Executed 1 job(s) for tool collection_creates_pair request (774.042 ms)

galaxy.tools.execute DEBUG 2020-05-29 18:36:35,855 [p:45925,w:1,m:0] [uWSGIWorker1Core1] Executed 1 job(s) for tool collection_creates_pair request (689.054 ms)

galaxy.tools.execute DEBUG 2020-05-29 18:37:27,540 [p:45925,w:1,m:0] [uWSGIWorker1Core1] Executed 1 job(s) for tool collection_creates_pair request (546.268 ms)

galaxy.tools.execute DEBUG 2020-05-29 18:37:50,067 [p:45925,w:1,m:0] [uWSGIWorker1Core1] Executed 1 job(s) for tool collection_creates_pair request (621.726 ms)

THIS BRANCH

galaxy.tools.execute DEBUG 2020-05-29 18:47:11,421 [p:46643,w:1,m:0] [uWSGIWorker1Core3] Executed 1 job(s) for tool collection_creates_pair request (400.876 ms)

galaxy.tools.execute DEBUG 2020-05-29 18:48:02,838 [p:46643,w:1,m:0] [uWSGIWorker1Core2] Executed 1 job(s) for tool collection_creates_pair request (384.065 ms)

galaxy.tools.execute DEBUG 2020-05-29 18:48:33,142 [p:46643,w:1,m:0] [uWSGIWorker1Core0] Executed 1 job(s) for tool collection_creates_pair request (424.079 ms)

galaxy.tools.execute DEBUG 2020-05-29 18:49:00,472 [p:46643,w:1,m:0] [uWSGIWorker1Core2] Executed 1 job(s) for tool collection_creates_pair request (450.527 ms)

galaxy.tools.execute DEBUG 2020-05-29 18:49:42,292 [p:46643,w:1,m:0] [uWSGIWorker1Core2] Executed 1 job(s) for tool create_2 request (324.669 ms)

galaxy.tools.execute DEBUG 2020-05-29 18:50:01,887 [p:46643,w:1,m:0] [uWSGIWorker1Core1] Executed 1 job(s) for tool create_2 request (349.538 ms)

galaxy.tools.execute DEBUG 2020-05-29 18:50:26,719 [p:46643,w:1,m:0] [uWSGIWorker1Core2] Executed 1 job(s) for tool collection_creates_pair request (380.079 ms)

galaxy.tools.execute DEBUG 2020-05-29 18:51:02,118 [p:46643,w:1,m:0] [uWSGIWorker1Core3] Executed 1 job(s) for tool collection_creates_pair request (378.154 ms)

galaxy.tools.execute DEBUG 2020-05-29 18:52:07,148 [p:46643,w:1,m:0] [uWSGIWorker1Core3] Executed 1 job(s) for tool create_2 request (379.535 ms)


DEV

galaxy.tools.execute DEBUG 2020-05-29 18:53:29,193 [p:48758,w:1,m:0] [uWSGIWorker1Core2] Executed 1 job(s) for tool create_2 request (418.890 ms)

galaxy.tools.execute DEBUG 2020-05-29 18:54:08,498 [p:48758,w:1,m:0] [uWSGIWorker1Core0] Executed 1 job(s) for tool create_2 request (342.142 ms)

galaxy.tools.execute DEBUG 2020-05-29 18:54:33,650 [p:48758,w:1,m:0] [uWSGIWorker1Core3] Executed 1 job(s) for tool create_2 request (369.027 ms)

galaxy.tools.execute DEBUG 2020-05-29 18:55:00,766 [p:48758,w:1,m:0] [uWSGIWorker1Core3] Executed 1 job(s) for tool collection_creates_pair request (562.289 ms)

galaxy.tools.execute DEBUG 2020-05-29 18:55:18,988 [p:48758,w:1,m:0] [uWSGIWorker1Core1] Executed 1 job(s) for tool collection_creates_pair request (510.890 ms)

galaxy.tools.execute DEBUG 2020-05-29 18:55:48,110 [p:48758,w:1,m:0] [uWSGIWorker1Core2] Executed 1 job(s) for tool collection_creates_pair request (468.725 ms)

galaxy.tools.execute DEBUG 2020-05-29 18:56:25,401 [p:48758,w:1,m:0] [uWSGIWorker1Core3] Executed 1 job(s) for tool collection_creates_pair request (527.128 ms)
```

### CODE REVIEW

File: lib/galaxy/managers/collections.py [hunk 1]
```diff
@@ -121,11 +122,13 @@ def create(self, trans, parent, name, collection_type, element_identifiers=None,
...
-        return self._create_instance_for_collection(
-            trans, parent, name, dataset_collection, implicit_inputs=implicit_inputs, implicit_output_name=implicit_output_name, tags=tags
+        rval = self._create_instance_for_collection(
+            trans, parent, name, dataset_collection, implicit_inputs=implicit_inputs, implicit_output_name=implicit_output_name, tags=tags, flush=flush, hid=hid
         )
+        return rval
 
-    def _create_instance_for_collection(self, trans, parent, name, dataset_collection, implicit_output_name=None, implicit_inputs=None, tags=None, flush=True):
+    def _create_instance_for_collection(self, trans, parent, name, dataset_collection, implicit_output_name=None, implicit_inputs=None, tags=None, flush=True, hid=None):
+        _create_instance_for_collection_timer = ExecutionTimer()
```
Comment (bgruening) [inline lib/galaxy/managers/collections.py]:
```suggestion

```

File: lib/galaxy/tools/actions/__init__.py [hunk 2]
```diff
@@ -496,10 +496,17 @@ def handle_output(name, output, hidden=None):
                         trans.sa_session.add(element)
                         current_element_identifiers.append({
                             "__object__": element,
+                            "__skip_object_session_check__": True,
                             "name": output_part_def.element_identifier,
                         })
-
-                    history.add_datasets(trans.sa_session, created_element_datasets, set_hid=set_output_hid, quota=False, flush=True)
+                    add_datasets = ExecutionTimer()
```
Comment (bgruening) [inline lib/galaxy/tools/actions/__init__.py]:
```suggestion

```

File: lib/galaxy/tools/actions/__init__.py [hunk 3]
```diff
@@ -496,10 +496,17 @@ def handle_output(name, output, hidden=None):
...
+                    reserve_hid_for_parent = output_collections.creating_hdca
+                    history.add_datasets(trans.sa_session, created_element_datasets, set_hid=set_output_hid, quota=False, flush=False, reserve_hid_for_parent=reserve_hid_for_parent)
+                    if reserve_hid_for_parent:
+                        hid = created_element_datasets[-1].hid + 1
+                    else:
+                        hid = None
+                    log.info("add_datasets %s" % add_datasets)
```
Comment (bgruening) [inline lib/galaxy/tools/actions/__init__.py]:
```suggestion

```

**Review by bgruening — COMMENTED** [review]
When the test are passing I can try to backport it to EU and test it extensively if you like.
Thanks John.

**Comment by jmchilton** [comment]
Ugh - I really messed up cherry picking the changes out of my branch with extra logging - sorry @bgruening

**Comment by bgruening** [comment]
Uh, green! @jmchilton should I give it a try on 19.09 EU?

**Comment by jmchilton** [comment]
Sounds good - let me know how it goes? Did that disabling that legacy setting help things at all?

**Comment by bgruening** [comment]
```
Jun 03 19:44:58 sn04.bi.uni-freiburg.de python[3074033]: galaxy.workflow.run DEBUG 2020-06-03 19:44:58,142 Marking step 758717 outputs of invocation 99366 delayed (dependent step [758716] delayed, so this step must be delayed)
Jun 03 19:44:58 sn04.bi.uni-freiburg.de python[3074033]: galaxy.workflow.run DEBUG 2020-06-03 19:44:58,144 Marking step 758718 outputs of invocation 99366 delayed (dependent step [758717] delayed, so this step must be delayed)
Jun 03 19:44:58 sn04.bi.uni-freiburg.de python[3074033]: galaxy.workflow.scheduling_manager DEBUG 2020-06-03 19:44:58,157 Workflow invocation [99366] scheduled
Jun 03 19:44:58 sn04.bi.uni-freiburg.de python[3074033]: galaxy.workflow.scheduling_manager DEBUG 2020-06-03 19:44:58,157 Attempting to schedule workflow invocation [(99367,)]
Jun 03 19:44:58 sn04.bi.uni-freiburg.de python[3074033]: galaxy.workflow.scheduling_manager DEBUG 2020-06-03 19:44:58,185 Attempting to schedule workflow invocation [(99429,)]
Jun 03 19:44:58 sn04.bi.uni-freiburg.de python[3073995]: galaxy.tools.actions INFO 2020-06-03 19:44:58,889 Handled collection output named output_paired_coll for tool toolshed.g2.bx.psu.edu/repos/iuc/fastp/fastp/0.20.1+galaxy0 (4700.853 ms)
Jun 03 19:44:58 sn04.bi.uni-freiburg.de python[3073995]: galaxy.tools.actions INFO 2020-06-03 19:44:58,952 Handled output named report_html for tool toolshed.g2.bx.psu.edu/repos/iuc/fastp/fastp/0.20.1+galaxy0 (62.085 ms)
Jun 03 19:44:58 sn04.bi.uni-freiburg.de python[3073995]: galaxy.tools.actions INFO 2020-06-03 19:44:58,996 Added output datasets to history (43.274 ms)
Jun 03 19:44:59 sn04.bi.uni-freiburg.de python[3073995]: galaxy.tools.actions INFO 2020-06-03 19:44:59,001 Setup for job Job[unflushed,tool_id=toolshed.g2.bx.psu.edu/repos/iuc/fastp/fastp/0.20.1+galaxy0] complete, ready to be enqueued (4.951 ms)
Jun 03 19:44:59 sn04.bi.uni-freiburg.de python[3073995]: galaxy.web_stack.handlers INFO 2020-06-03 19:44:59,086 Flushed transaction for Job[id=8549119,tool_id=toolshed.g2.bx.psu.edu/repos/iuc/fastp/fastp/0.20.1+galaxy0] (84.940 ms)
Jun 03 19:44:59 sn04.bi.uni-freiburg.de python[3073995]: galaxy.web_stack.handlers INFO 2020-06-03 19:44:59,086 (Job[id=8549119,tool_id=toolshed.g2.bx.psu.edu/repos/iuc/fastp/fastp/0.20.1+galaxy0]) Handler '_default_' assigned using 'db-skip-locked' assignment method
Jun 03 19:44:59 sn04.bi.uni-freiburg.de python[3073995]: galaxy.tools.execute DEBUG 2020-06-03 19:44:59,087 Tool [toolshed.g2.bx.psu.edu/repos/iuc/fastp/fastp/0.20.1+galaxy0] created job [8549119] (4955.356 ms)
Jun 03 19:45:00 sn04.bi.uni-freiburg.de python[3074033]: galaxy.workflow.run DEBUG 2020-06-03 19:45:00,399 Marking step 273942 outputs of invocation 99429 delayed (dependent collection [285700] not yet populated with datasets)
Jun 03 19:45:00 sn04.bi.uni-freiburg.de python[3074033]: galaxy.workflow.run DEBUG 2020-06-03 19:45:00,400 Marking step 273943 outputs of invocation 99429 delayed (dependent collection [285700] not yet populated with datasets)
```

Unfortunately, this does not look better, but maybe it is 19.09 that does not like the patch.

> Did that disabling that legacy setting help things at all?

Yes, but I think in different ways. Zergling pressure is less and our uwsgi-queue-full-errors seem to be less, subjectively.

**Pull request closed without merging**